### DISCUSSION TITLE
HTML to PDF Plugin in `PDF-Package`

### DISCUSSION THREAD

**Discussion by pritamps** [Samagra-Development/SamagraX-community#39] [General Discussions]
I have started work on a HTML to PDF transformer to enable more flexible creation of templates for various use cases.

An initial thought was to try it with Google Doc's HTML export function. Currently, the Google Docs Plugin (and ODK plugin) need to connect to an Apps Script to fill in template variables and generate the PDF. Downloading the doc as a HTML would allow all this to be done on the server itself. 

As a POC, I used python's `pdfkit` library (which uses `wkhtmltopdf` as its backend). However, this did not work too well because:

1. The HTML is very dirty. Can still find and replace the <<x>> though, so it might still work out
2. Page breaks don't appear in the HTML naturally. I thought we could deal with this by adding a <<NewPage>>. Kind of hacky, but could work
3. The headers and footers don't carry over to multiple pages. Header is shown only on the first page, and footer is shown only on the last page.
4. Alignment of images, tables is not preserved.

As an illustration, here is the doc generated from the template:

[out.pdf](https://github.com/Samagra-Development/SamagraX-community/files/6690391/out.pdf)

With this information, I don't think this will work, and we will need to create a HTML templating engine for this effort to be worth it. One possibility is to use the [`Jinja2`](https://pypi.org/project/Jinja2/) as the templating engine. 

ls Is this something the UCI team will want to consider?

**Comment by aayushkedia95** [comment]
@ChakshuGautam, can you help here?
